File: dart/test/walker_test.dart
import 'package:dart/walker.dart';
import 'package:dart/point.dart';
import 'package:test/test.dart';
import 'dart:math';

void main() {
	test('no marbles', () {
		final rng = Random();
		final w = Walker(rng, 0, 5);
		expect(w.steps(), 1);
		expect(w.intersections(), 0);
	});

	test('no steps', () {
		final rng = Random();
		final w = Walker(rng, 4, 0);
		expect(w.steps(), 1);
		expect(w.intersections(), 0);
	});

	test('double zero', () {
		final rng = Random();
		final w = Walker(rng, 0, 0);
		expect(w.steps(), 1);
		expect(w.intersections(), 0);
	});

	test('marble placement', () {
		final rng = Random();
		final w = Walker(rng, 4, 1);
		expect(w.marbles.length, 1);
		final hasEvenParity = (Point p) => (p.x + p.y) % 2 == 0;
		expect(hasEvenParity(w.marbles[0]), isTrue);
	});

	test('normal inputs', () {
		final rng = Random();
		final w = Walker(rng, 3, 5);
		expect(w.steps(), greaterThanOrEqualTo(8));
		expect(w.steps(), lessThanOrEqualTo(16));
		expect(w.intersections(), greaterThanOrEqualTo(2));
		expect(w.intersections(), lessThanOrEqualTo(5));
	});
}

File: cpp/walker_test.cpp
#include "walker.hpp" // Walker
#include <cassert>    // assert
#include <iostream>   // cout

using namespace std;

int main() {
	// No marbles
	{
		Walker w(0, 5);
		assert(w.steps() == 1);
		assert(w.intersections() == 0);
	}
	// No steps
	{
		Walker w(4, 0);
		assert(w.steps() == 1);
		assert(w.intersections() == 0);
	}
	// Double zero
	{
		Walker w(0, 0);
		assert(w.steps() == 1);
		assert(w.intersections() == 0);
	}
	// Marble placement
	{
		Walker w(4, 1);
		assert(w.firstMarbleParity());
	}
	// Normal inputs
	{
		Walker w(3, 5);
		assert(w.steps() >= 8);
		assert(w.steps() <= 16);
		assert(w.intersections() >= 2);
		assert(w.intersections() <= 5);
	}
	cout << "Test passed.\n";
	return 0;
}
